File: Frontend/src/app/product/models/product.ts
import {Category} from '../../category/models/category';

export class Product {
  private _id: string = "";
  private _name: string = "";
  private _category: Category = new Category("","");
  private _price: number = 0;
  private _description: string = "";
  private _imageName: string = "";

  public constructor(id: string, name: string, category: Category, price: number, description: string, imageName: string) {
    this._id = id;
    this._name = name;
    this._category = category;
    this._price = price;
    this._description = description;
    this._imageName = imageName;
  }

  public get id(): string {
    return this._id;
  }

  public set id(value: string) {
    this._id = value;
  }

  public get name(): string {
    return this._name;
  }

  public set name(name: string) {
    this._name = name;
  }

  public get category(): Category {
    return this._category;
  }

  public set category(category: Category)  {
    this._category = category;
  }

  public get price(): number {
    return this._price;
  }

  public set price(price: number) {
    this._price = price;
  }

  public get description(): string {
    return this._description;
  }

  public set description(description: string) {
    this._description = description;
  }

  public get imageName(): string {
    return this._imageName;
  }

  public set imageName(value: string) {
    this._imageName = value;
  }

  public toJSON(): any {
    return {
      id: this._id,
      name: this._name,
      category: this._category instanceof Category ? this._category.toJSON() : this._category,
      price: this._price,
      description: this._description,
      imageName: this._imageName
    };
  }
}

File: Backend/src/main/java/iprwc/backend/models/Product.java
package iprwc.backend.models;

import jakarta.persistence.*;
import org.hibernate.annotations.JdbcTypeCode;
import java.sql.Types;
import java.util.UUID;

@Entity
@Table(name = "Product")
public class Product {

    @jakarta.persistence.Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @JdbcTypeCode(Types.VARCHAR)
    private UUID id;

    private String name;

    private double price;

    private String description;

    @ManyToOne
    private Category category;

    private String imageName;

    public Product() { }

    public Product(String name, double price, String description, Category category, String imageName) {
        this.name = name;
        this.price = price;
        this.description = description;
        this.category = category;
        this.imageName = imageName;
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public Category getCategory() {
        return category;
    }

    public void setCategory(Category category) {
        this.category = category;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getImageName() {
        return imageName;
    }

    public void setImageName(String imageName) {
        this.imageName = imageName;
    }

    @Override
    public String toString() {
        return "Product{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", price=" + price +
                ", description='" + description + '\'' +
                ", category=" + category +
                ", imageName='" + imageName + '\'' +
                '}';
    }
}
